retrieval/aggregator_test.go

```diff
@@ -0,0 +1,84 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package retrieval
+
+import (
+	"bytes"
+	"context"
+	"math"
+	"reflect"
+	"testing"
+
+	"github.com/go-kit/kit/log"
+	promlabels "github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/tsdb/labels"
+	"go.opencensus.io/stats"
+)
+
+func TestCounterAggregator(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	type input struct {
+		ts    int64
+		value float64
+	}
+
+	for _, tt := range []struct {
+		name  string
+		input []input
+		want  []float64
+	}{
+		{"simple", []input{input{1, 15}, input{2, 25}}, []float64{10}},
+		{"counter resets", []input{input{1, 15}, input{2, 25}, input{3, 5}, input{4, 25}, input{5, 15}}, []float64{10, 5, 20, 15}},
+		{"NaNs are ignored", []input{input{1, 15}, input{2, math.NaN()}, input{3, 25}}, []float64{10}},
+		{"out of order points are ignored", []input{input{1, 15}, input{3, 25}, {2, 20}, {4, 30}}, []float64{10, 5}},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			logBuffer := &bytes.Buffer{}
+			defer func() {
+				if logBuffer.Len() > 0 {
+					t.Log(logBuffer.String())
+				}
+			}()
+			logger := log.NewLogfmtLogger(logBuffer)
+
+			var got []float64
+			statsRecord = func(ctx context.Context, ms ...stats.Measurement) {
+				for _, m := range ms {
+					got = append(got, m.Value())
+				}
+			}
+			defer func() { statsRecord = stats.Record }()
```

Comment:
I'm not too excited about replacing a global function in a different file, because we can accidentally end up with a bug that's very hard to understand. How about injecting the dependency directly in the counter aggregator instance below?

Reply:
Done.